sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushLocalTopKThroughOuterJoin.scala

```diff
@@ -0,0 +1,118 @@
+import org.apache.spark.sql.catalyst.trees.TreePattern.{LIMIT, OUTER_JOIN, SORT}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This rule supports push down local limit and local sort from TopK through other join:
+ *   - for a left outer join, the references of ordering of TopK come from the left side and
+ *     the limits of TopK is smaller than left side max rows
+ *   - for a right outer join, the references of ordering of TopK come from the right side and
+ *     the limits of TopK is smaller than right side max rows
+ */
+object PushLocalTopKThroughOuterJoin extends Rule[LogicalPlan] {
+  private def smallThan(limits: Int, maxRowsOpt: Option[Long]): Boolean = maxRowsOpt match {
+    case Some(maxRows) => limits < maxRows
+    case _ => true
+  }
+
+  private def canPushThroughOuterJoin(
+      joinType: JoinType,
+      order: Seq[SortOrder],
+      leftChild: LogicalPlan,
+      rightChild: LogicalPlan,
+      limits: Int): Boolean = joinType match {
+    case LeftOuter =>
+      order.forall(_.references.subsetOf(leftChild.outputSet)) &&
+        smallThan(limits, leftChild.maxRowsPerPartition)
+    case RightOuter =>
+      order.forall(_.references.subsetOf(rightChild.outputSet)) &&
+        smallThan(limits, rightChild.maxRowsPerPartition)
+    case _ => false
+  }
+
+  private def findOuterJoin(limits: Int, order: Seq[SortOrder], child: LogicalPlan): Seq[Join] = {
+    child match {
+      case j @ ExtractEquiJoinKeys(joinType, _, _, _, _, leftChild, rightChild, _)
+          if canPushThroughOuterJoin(joinType, order, leftChild, rightChild, limits) =>
+        // we should find the bottom outer join which can push local topK
+        val childOuterJoin = joinType match {
+          case LeftOuter => findOuterJoin(limits, order, leftChild)
+          case RightOuter => findOuterJoin(limits, order, rightChild)
+          case _ => Seq.empty
+        }
+        if (childOuterJoin.nonEmpty) {
+          childOuterJoin
+        } else {
+          j :: Nil
+        }
+      case u: Union => u.children.flatMap(child => findOuterJoin(limits, order, child))
+      case p: Project if p.projectList.forall(_.deterministic) =>
+        findOuterJoin(limits, order, p.child)
+      case f: Filter if f.condition.deterministic =>
+        findOuterJoin(limits, order, f.child)
+      case r: RepartitionByExpression if r.partitionExpressions.forall(_.deterministic) =>
+        findOuterJoin(limits, order, r.child)
+      case r: RebalancePartitions if r.partitionExpressions.forall(_.deterministic) =>
+        findOuterJoin(limits, order, r.child)
+      case r: Repartition => findOuterJoin(limits, order, r.child)
+      case _ => Seq.empty
+    }
+  }
+
+  private def pushLocalTopK(limits: Int, order: Seq[SortOrder], join: Join): Join = {
+    val (newLeft, newRight) = join.joinType match {
+      case LeftOuter =>
+        (LocalLimit(Literal(limits), Sort(order, false, join.left)), join.right)
```

Comment:
Could we put this logic into `LimitPushDown`?

Reply:
this new rule is not simple enough, so I prefer to be independent from LimitPushDown. have other concern about merging into LimitPushDown ?